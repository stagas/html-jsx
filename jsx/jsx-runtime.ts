import type * as jsxi from '../types'

declare global {
  namespace JSX {
    /** The type returned by the `jsx` factory. */
    type Element = VKid

    interface IntrinsicElements extends jsxi.IntrinsicElements {
      [k: string]: unknown
    }

    interface IntrinsicAttributes {
      /** List index key - each item's `key` must be unique. */
      key?: string | number

      /** Passed from above children of this component. */
      children?: VKid
    }

    interface HTMLAttributes<T> extends jsxi.HTMLAttributes<T> {}
    interface SVGAttributes<T> extends jsxi.SVGAttributes<T> {}
    interface DOMAttributes<T> extends jsxi.DOMAttributes<T> {}
  }
}

export type $<T> = (props: T & { children?: any }) => JSX.Element

export const Fragment = ({ children }: any) => children

export const jsx = (tag: VNode['tag'], props: VNode['props'], key: VNode['key']): VNode => ({
  tag,
  props,
  key,
  toString,
})

export const jsxs = jsx

export interface VNode {
  tag: string | typeof Fragment
  props: Record<string, any>
  key: string | number | undefined | void
  toString(): string
}

export type VKid = VKid[] | VNode | string | number | boolean | null | undefined | void

// const trim = (s: string) => s.trim().replace(/(>)(\s{2,})|(\s{2,})(<\/)/g, '$1$4')
const join = (c: string | string[]) => (Array.isArray(c) ? c : c == null ? [] : [c]).map(x => '' + x).join('  ')
const toString = function(this: VNode) {
  const { tag, props } = this
  if (typeof tag != 'string') return join(tag(props))
  const attrs = Object.entries(props)
    .filter(([key]) => !['children'].includes(key))
    .map(([key, value]) =>
      typeof value == 'boolean'
        ? value
          ? key
          : ''
        : `${key}="${
          typeof value == 'function'
            ? `jsxHandlers.get(${Id})[${H.push(value) - 1}].call(this, event)`
            : ('' + value).replace(/&/g, '&amp;').replace(/"/g, '&quot;')
        }"`
    )
  return `<${tag}${attrs.length ? ' ' + attrs.join(' ') : ''}>${join(props.children)}</${tag}>`
}

let Id = 0
const ids = new WeakMap()

const handlers = ((globalThis as any).jsxHandlers ??= new Map<number, any>([[0, []]]))
let H: any = handlers.get(0)
const registry = new FinalizationRegistry((id: string) => handlers.delete(+id))

export const render = (jsx: VKid, el: Element | ShadowRoot) => {
  let id = ids.get(el)
  if (!id) ids.set(el, id = (Math.random() * performance.now() * 1e3) | 0)
  handlers.set(Id = id, H = [])
  registry.register(el, id)
  el.innerHTML = '' + (jsx ?? '')
  H = handlers.get(Id = 0)
}
