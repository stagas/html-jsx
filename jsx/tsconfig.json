{
  "extends": "../tsconfig.dist.json",
  "include": [
    "."
  ]
}
