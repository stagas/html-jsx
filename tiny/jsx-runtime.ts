import type * as jsxi from '../types'

declare global {
  namespace JSX {
    /** The type returned by the `jsx` factory. */
    type Element = string

    interface IntrinsicElements extends jsxi.IntrinsicElements {
      [k: string]: unknown
    }

    interface IntrinsicAttributes {
      /** List index key - each item's `key` must be unique. */
      key?: string | number

      /** Passed from above children of this component. */
      children?: string
    }

    interface HTMLAttributes<T> extends jsxi.HTMLAttributes<T> {}
    interface SVGAttributes<T> extends jsxi.SVGAttributes<T> {}
    interface DOMAttributes<T> extends jsxi.DOMAttributes<T> {}
  }
}

export type $<T> = (props: T & { children?: any }) => JSX.Element

export const Fragment = ({ children }: any) => children

export const jsx = (tag: VNode['tag'], props: VNode['props'], _key: VNode['key']): string => {
  if (typeof tag != 'string') return join(tag(props))
  const attrs = Object.entries(props)
    .filter(([key]) => !['children'].includes(key))
    .map(([key, value]) =>
      typeof value == 'boolean'
        ? value
          ? key
          : ''
        : `${key}="${('' + value).replace(/&/g, '&amp;').replace(/"/g, '&quot;')}"`
    )
  return `<${tag}${attrs.length ? ' ' + attrs.join(' ') : ''}>${join(props.children)}</${tag}>`
}

export const jsxs = jsx

export interface VNode {
  tag: string | typeof Fragment
  props: Record<string, any>
  key: string | number | undefined | void
  toString(): string
}

export type VKid = VKid[] | VNode | string | number | boolean | null | undefined | void

// const trim = (s: string) => s.trim().replace(/(>)(\s{2,})|(\s{2,})(<\/)/g, '$1$4')
const join = (c: string | string[]) => (Array.isArray(c) ? c : c == null ? [] : [c]).join('  ')

export const render = (jsx: VKid, el: Element | ShadowRoot) => {
  el.innerHTML = (jsx as string) ?? ''
}
