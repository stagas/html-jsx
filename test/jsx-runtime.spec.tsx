/** @jsxImportSource ../jsx */
import { render } from '../jsx/jsx-runtime'

declare global {
  let jsxHandlers: any
}

const r = (jsx: any) => {
  const div = document.createElement('div')
  render(jsx, div)
  return div
}

describe('html-jsx', () => {
  it('renders jsx to text', () => {
    expect('' + <div>hello world</div>).toMatchSnapshot()
  })

  it('renders attributes', () => {
    expect('' + <input type="text" value={'hello w"orld'} />).toMatchSnapshot()
  })

  it('fragment', () => {
    expect(
      ''
        + (
          <>
            <div>hello world</div>
            <p>a paragraph</p>
          </>
        )
    ).toMatchSnapshot()
  })

  it('component', () => {
    const Foo = ({ label }: any) => <label>{label}</label>
    expect(
      r(
        <>
          <Foo label="foo" />
        </>
      ).innerHTML
    ).toMatchSnapshot()
  })

  it('list', () => {
    expect(
      ''
        + (
          <ul>
            {[1, 2, 3].map(i => <li>{i}</li>)}
          </ul>
        )
    ).toMatchSnapshot()
  })

  it('handler', () => {
    let count = 0
    const div = r(<button onclick={() => count++}>click me</button>) as any
    div.firstChild.click()
    expect(count).toBe(1)
  })

  it('null', () => {
    expect(r(null).innerHTML).toBe('')
  })

  it('check gc', async () => {
    expect(jsxHandlers.size).toBeGreaterThan(1)
    gc!()
    await new Promise(resolve => setTimeout(resolve, 10))
    expect(jsxHandlers.size).toBe(1)
  })
})
