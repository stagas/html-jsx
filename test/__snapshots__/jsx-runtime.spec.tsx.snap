// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html-jsx component 1`] = `"<label>foo</label>"`;

exports[`html-jsx fragment 1`] = `"<div>hello world</div>  <p>a paragraph</p>"`;

exports[`html-jsx list 1`] = `"<ul><li>1</li>  <li>2</li>  <li>3</li></ul>"`;

exports[`html-jsx renders attributes 1`] = `"<input type=\\"text\\" value=\\"hello w&quot;orld\\"></input>"`;

exports[`html-jsx renders jsx to text 1`] = `"<div>hello world</div>"`;
